### DISCUSSION TITLE
Union of tuples with literal types doesn't narrow as expected

### DISCUSSION THREAD

**Discussion by bwburnsides** [microsoft/pyright#6686] [Q&A] [answered]
Hello!

I am working on a bootstrap compiler in Python 3.12 and relying on pyright for static type analysis. I've just ran into an issue where the inferred type of a match target is not what I expect it to be. Here's my setup:
```python
class ResultKind(enum.Enum):
    Match = enum.auto()
    NoMatch = enum.auto()
    Error = enum.auto()

class ParseError:
    ...

type ParseResult[T] = (
    Lit[ResultKind.NoMatch]
    | tuple[Lit[ResultKind.Match], T]
    | tuple[Lit[ResultKind.Error], ParseError]
)

def test_parse() -> ParseResult[int]:
    ...

match test_parse():
    case ResultKind.NoMatch:
        ...
    case (ResultKind.Match, item):
        ...
    case (ResultKind.Error, err):
        ...
```
In the `(ResultKind.Match, item)` case, I expect `item` to be of type `int`. However, pyright finds it to be `int | ParseError`. And indeed if I swap the last two cases, it'll find `err` to be of type `int | ParseError`. Given the type alias explicitly mapping `ResultKind.Match` to `T`, and `ParseKind.Error` to `ParseError`, this is unexpected to me.

If I only include the `ResultKind.NoMatch` case, then pyright gives the warning I expect, which seems to contrast with the behavior show above.
```
Unhandled type: "tuple[Literal[ResultKind.Match], int] | tuple[Literal[ResultKind.Error], ParseError]"
```

What am I missing here?

**Comment by erictraut** [comment]
Yes, this looks wrong. I'm going to create a bug report from this discussion topic and investigate further.

**Reply by erictraut** [reply, marked as answer]
Here's a link to the [bug report](https://github.com/microsoft/pyright/issues/6691), for reference.

**Reply by bwburnsides** [reply]
Excellent, I'll keep an eye on it. Thanks for the response! I'm glad I wasn't missing something obvious.
